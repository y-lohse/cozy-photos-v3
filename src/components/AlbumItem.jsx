/* global cozy */
import styles from '../styles/albumsList'
import classNames from 'classnames'

import React, { Component } from 'react'
import { Link, withRouter } from 'react-router'
import { translate } from 'cozy-ui/react/I18n'

import { fetchAlbumCover } from '../ducks/albums'
import ImageLoader from './ImageLoader'

const isAlbumEmpty = album => !(album && album.photos && album.photos.length !== 0)

const SharedIcon = ({ byMe }) => (
  <div className={styles['pho-album-shared']}>
    <div className={classNames(styles['pho-album-shared-icon'], styles[byMe ? '--by-me' : '--with-me'])} />
  </div>
)

export class AlbumItem extends Component {
  constructor (props) {
    super(props)
    this.state = {
      isLoading: true,
      coverPhoto: null
    }
  }

  componentDidMount () {
    const { album } = this.props
    if (isAlbumEmpty(album)) {
      return this.setState({ isLoading: false })
    }
    fetchAlbumCover(album)
      .then(photo => {
        this.setState({ coverPhoto: photo, isLoading: false })
      })
      .catch(error => {
        this.setState({ isLoading: false })
        console.log(error)
      })
  }

  render () {
    if (this.state.isLoading) {
      return null
    }
    const { t, album, sharedByMe, sharedWithMe, onClick } = this.props
    const { coverPhoto } = this.state

    const image = !coverPhoto
      ? <div
        className={styles['pho-album-photo-item']}
      />
      : <ImageLoader
        className={styles['pho-album-photo-item']}
        alt={`${album.name} album cover`}
        photo={coverPhoto}
        src={`${cozy.client._url}${coverPhoto.links.small}`}
      />
    const desc = <h4 className={styles['pho-album-description']}>
      {t('Albums.album_item_description',
        {smart_count: album.photos.length})
      }
      {(sharedByMe || sharedWithMe) && ` - ${t('Albums.album_item_shared_ro')}`}
    </h4>
    const title = <h2 className={styles['pho-album-title']}>{album.name}</h2>

    if (onClick) {
      return (
        <div className={styles['pho-album']}>
          <div onClick={() => onClick(album)} className={styles['pho-album-link']}>
            {image}{title}{desc}
          </div>
          {(sharedByMe || sharedWithMe) && <SharedIcon byMe={false} />}
        </div>
      )
    }
    const parentPath = this.props.router.location.pathname
    return (
      <div className={styles['pho-album']}>
        <Link to={`${parentPath}/${album._id}`} className={styles['pho-album-link']}>
          {image}{title}{desc}
        </Link>
        {(sharedByMe || sharedWithMe) && <SharedIcon byMe={sharedByMe} />}
      </div>
    )
  }
}

export default translate()(withRouter(AlbumItem))
