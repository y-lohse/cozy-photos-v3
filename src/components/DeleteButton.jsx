import React from 'react'
import { translate } from 'cozy-ui/react/I18n'

const styles = {
  parent: {
    position: 'relative',
    overflow: 'hidden'
  },
  input: {
    position: 'absolute',
    top: 0,
    left: 0,
    opacity: 0,
    width: '100%',
    height: '100%',
    zIndex: 1
  }
}

export const DeleteButton = ({ t, label, type = 'button', disabled, onDelete, className = '' }) => (
  <label
    role='button'
    disabled={disabled}
    className={`${className} ${type === 'menu-item' ? 'coz-link--delete' : 'coz-btn coz-btn--danger coz-btn--delete'}`}
    style={styles.parent}
  >
    { label }
    <button
      disabled={disabled}
      style={styles.input}
      onClick={() => onDelete()}
    />
  </label>
)

export default translate()(DeleteButton)
