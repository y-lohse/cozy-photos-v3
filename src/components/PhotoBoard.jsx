import styles from '../styles/photoList'

import React, { Component } from 'react'
import AutoSizer from 'react-virtualized/dist/commonjs/AutoSizer'
import { translate } from 'cozy-ui/react/I18n'

import PhotoList from './PhotoList'
import Empty from './Empty'
import Loading from './Loading'
import ErrorComponent from './ErrorComponent'

const Spinner = () => <div class={styles['pho-list-spinner']} />

class MoreButton extends Component {
  constructor (props) {
    super(props)
    this.state = {
      fetching: false
    }
  }

  handleClick () {
    this.setState({ fetching: true })
    this.props.onClick()
      .then(() => this.setState({ fetching: false }))
  }

  render () {
    const { children, width } = this.props
    if (!width) {
      return null
    }
    const { fetching } = this.state
    return (
      <div style={{ width: width }} className={styles['pho-list-morebutton']}>
        {fetching &&
          <button className='coz-btn' disabled>
            <Spinner />
          </button>
        }
        {!fetching &&
          <button
            className='coz-btn coz-btn--secondary'
            onClick={() => this.handleClick()}
          >
            {children}
          </button>
        }
      </div>
    )
  }
}

export class PhotoBoard extends Component {
  render () {
    const {
      t,
      lists,
      selected,
      photosContext,
      showSelection,
      onPhotoToggle,
      onPhotosSelect,
      onPhotosUnselect,
      fetchStatus,
      hasMore,
      onFetchMore
    } = this.props

    const isError = fetchStatus === 'failed'
    const isFetching = fetchStatus === 'pending' || fetchStatus === 'loading'

    if (isError) {
      return <ErrorComponent errorType={`${photosContext}_photos`} />
    }
    if (isFetching) {
      return <Loading loadingType='photos_fetching' />
    }
    if (!isFetching && (lists.length === 0 || lists[0].photos.length === 0)) {
      return <Empty emptyType={`${photosContext}_photos`} />
    }

    return (
      <AutoSizer>
        {({ width, height }) => (
          <div className={showSelection ? styles['pho-list-selection'] : ''}>
            {lists.map(photoList =>
              <PhotoList
                key={photoList.title}
                title={photoList.title}
                photos={photoList.photos}
                selected={selected}
                showSelection={showSelection}
                onPhotoToggle={onPhotoToggle}
                onPhotosSelect={onPhotosSelect}
                onPhotosUnselect={onPhotosUnselect}
                containerWidth={width}
              />
            )}
            {hasMore &&
              <MoreButton width={width} onClick={onFetchMore}>
                {t('Board.load_more')}
              </MoreButton>
            }
          </div>
        )}
      </AutoSizer>
    )
  }
}

export default translate()(PhotoBoard)
