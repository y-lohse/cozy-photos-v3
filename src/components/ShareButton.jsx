import React from 'react'

import classNames from 'classnames'

export const ShareButton = ({ label, onClick }) => (
  <button
    role='button'
    className={classNames('coz-btn', 'coz-btn--secondary', 'coz-btn--share')}
    onClick={() => onClick()}
  >
    {label}
  </button>
)

export default ShareButton
