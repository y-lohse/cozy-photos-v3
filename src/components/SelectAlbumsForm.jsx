import styles from '../styles/albumsList'

import React from 'react'
import classNames from 'classnames'

import { withError } from '../components/ErrorComponent'
import { withEmpty } from '../components/Empty'

import AlbumItem from '../components/AlbumItem'

const DumbAlbumsList = props => (
  <div className={classNames(styles['pho-album-list'], styles['pho-album-list--thumbnails'], styles['pho-album-list--selectable'])}>
    {props.albums.entries.map((a) => <AlbumItem album={a} key={a._id} onServerError={props.onServerError} onClick={props.onSubmitSelectedAlbum} />)}
  </div>
)

const AlbumsList = withEmpty(props => props.albums.entries.length === 0, 'albums', DumbAlbumsList)

const DumbAlbumsView = props => (
  <div>
    <AlbumsList {...props} />
  </div>
)

const ErrorAlbumsView = withError(props => props.error, 'albums', DumbAlbumsView)

export const AlbumsView = (props) => {
  if (!props.albums) {
    return null
  }
  const error = props.albums.fetchingStatus === 'failed'
  return <ErrorAlbumsView error={error} onServerError={() => this.handleError(error)} {...props} />
}

export default AlbumsView
