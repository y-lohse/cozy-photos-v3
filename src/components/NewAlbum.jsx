import styles from '../styles/newAlbum'

import React, { Component } from 'react'
import classNames from 'classnames'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'

import { translate } from 'cozy-ui/react/I18n'
import { fetchIfNeededPhotos, fetchMorePhotos, getTimelineList, getPhotosByMonth, refetchSomePhotos } from '../ducks/timeline'
import { checkUniquenessOfAlbumName, createAlbum } from '../ducks/albums'
import PhotoBoard from './PhotoBoard'
import Alerter from './Alerter'

class NewAlbum extends Component {
  constructor (props) {
    super(props)
    this.state = {
      selected: [],
      name: props.t('Albums.create.panel_form.placeholder')
    }
  }

  onPhotoToggle = obj => {
    this.setState(({ selected }) => {
      const idx = selected.findIndex(i => i === obj.id)
      return {
        selected: idx === -1
          ? [...selected, obj.id]
          : [...selected.slice(0, idx), ...selected.slice(idx + 1)]
      }
    })
  }

  onPhotosSelect = ids => {
    this.setState(({ selected }) => {
      const newIds = ids.filter(id => selected.indexOf(id) === -1)
      return {
        selected: [...selected, ...newIds]
      }
    })
  }

  onPhotosUnselect = ids => {
    this.setState(({ selected }) => {
      return {
        selected: selected.filter(id => ids.indexOf(id) === -1)
      }
    })
  }

  onNameChange = e => {
    this.setState({ name: e.target.value })
  }

  onSubmit = async (e) => {
    try {
      const { name, selected } = this.state
      if (!name) {
        Alerter.error('Albums.create.error.name_missing')
        this.input.focus()
        return
      }
      // TODO: we should not have to dispatch this method (see comments in redux-cozy-api)
      const unique = await this.props.checkUniquenessOfAlbumName(name)
      if (!unique) {
        Alerter.error('Albums.create.error.already_exists', { name })
        this.input.focus()
        this.input.select()
        return
      }
      const album = await this.props.createAlbum(name, selected)
      // TODO: sadly we need to refetch the photos so that their relationships
      // property get updated and so that isRelated works when deleting a photo
      // that has just been added to an album
      await this.props.refetchSomePhotos(selected)
      this.props.closeAddAlbum()
      Alerter.success('Albums.create.success', {name: name, smart_count: selected.length})
      this.props.router.push(`/albums/${album.id}`)
    } catch (error) {
      Alerter.error('Albums.create.error.generic')
      console.log(error)
    }
  }

  componentDidMount () {
    this.props.fetchIfNeededPhotos()
    this.input.focus()
    this.input.select()
  }

  render () {
    const { t, closeAddAlbum } = this.props
    const { name } = this.state
    return (
      <div className={styles['pho-panel']}>
        <div className={styles['pho-panel-form']}>
          <header className={styles['pho-panel-header']}>
            <div className={styles['pho-panel-wrap']}>
              <label className={styles['coz-form-label']}>{t('Albums.create.panel_form.label')}</label>
              <input
                type='text'
                ref={input => { this.input = input }}
                value={name}
                onChange={this.onNameChange}
              />
            </div>
          </header>
          <div className={styles['pho-panel-content']}>
            <div className={styles['pho-panel-wrap']}>
              {this.renderBoard()}
            </div>
          </div>
          <footer className={styles['pho-panel-footer']}>
            <div className={styles['pho-panel-wrap']}>
              <div className={styles['pho-panel-controls']}>
                <button className={classNames('coz-btn', 'coz-btn--secondary')} onclick={closeAddAlbum}>
                  {t('Albums.create.panel_form.cancel')}
                </button>
                <button
                  className={classNames('coz-btn', 'coz-btn--regular')}
                  onClick={this.onSubmit}
                >
                  {t('Albums.create.panel_form.submit')}
                </button>
              </div>
            </div>
          </footer>
        </div>
      </div>
    )
  }

  renderBoard () {
    const { f, list, fetchMorePhotos } = this.props
    if (!list) {
      return null
    }
    const photoLists = getPhotosByMonth(list.entries, f, 'MMMM YYYY')
    return (
      <PhotoBoard
        lists={photoLists}
        selected={this.state.selected}
        photosContext='albums'
        showSelection
        onPhotoToggle={this.onPhotoToggle}
        onPhotosSelect={this.onPhotosSelect}
        onPhotosUnselect={this.onPhotosUnselect}
        fetchStatus={list.fetchStatus}
        hasMore={list.hasMore}
        onFetchMore={() => fetchMorePhotos(list.index, list.entries.length)}
      />
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  list: getTimelineList(state)
})

export const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchIfNeededPhotos: () => dispatch(fetchIfNeededPhotos()),
  fetchMorePhotos: (index, skip) => dispatch(fetchMorePhotos(index, skip)),
  createAlbum: (name, photos) => dispatch(createAlbum(name, photos)),
  // TODO: we should not have to dispatch this method (see comments in redux-cozy-api)
  checkUniquenessOfAlbumName: (name) => dispatch(checkUniquenessOfAlbumName(name)),
  refetchSomePhotos: (photos) => dispatch(refetchSomePhotos(photos))
})

export default withRouter(translate()(connect(
  mapStateToProps,
  mapDispatchToProps
)(NewAlbum)))
