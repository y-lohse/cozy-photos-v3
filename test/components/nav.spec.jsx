'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { mockT } from 'cozy-ui/react/I18n'
import { Nav } from '../../src/components/Nav'

describe('Nav component', () => {
  it('should be rendered correctly', () => {
    const component = shallow(
      <Nav t={mockT} />
    ).node
    expect(component).toMatchSnapshot()
  })
})
