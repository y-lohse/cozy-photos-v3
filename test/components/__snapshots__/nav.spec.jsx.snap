exports[`Nav component should be rendered correctly 1`] = `
<nav>
  <ul
    className="coz-nav">
    <li
      className="coz-nav-item">
      <Link
        activeClassName="active"
        className="pho-cat-photos coz-nav-link"
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/photos">
        Photos
      </Link>
    </li>
    <li
      className="coz-nav-item">
      <Link
        activeClassName="active"
        className="pho-cat-albums coz-nav-link"
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/albums">
        Albums
      </Link>
    </li>
  </ul>
</nav>
`;
